### ISSUE TITLE
Partial time derivatives with functions returning VectorValue 

### ISSUE THREAD

**Issue by santiagobadia** [gridap/GridapODEs.jl#14]
@fverdugo can you tell me whether you find a way to implement **partial derivatives** using `ForwardDiff`. I think I have tried all the things that have come to my mind without success

```julia
# ForwardDiff for multivariate functions

# The case of univariate function is clear, as already implemented in Gridap
h(x) = VectorValue(5*x[1]*x[2],x[2]^2)
dh(x) = ForwardDiff.jacobian(y->h(y).array,x.array)
dh(x) = ForwardDiff.jacobian(y->h(y).array,x.array)
dh(xv)

# As soon as H does not depend on t it is OK
H(x,t) = VectorValue(5*x[1]*x[2],x[2]^2)
Ht(t) = x -> H(x,t)
Hx(x) = t -> H(x,t)
dH(t) = x -> ForwardDiff.derivative(y->Hx(x)(y).array,t)
dH(tv)(xv)

# But what I would expect to work for multivariate functions does not
H(x,t) = VectorValue(5*x[1]*x[2],x[2]^2+t)
Ht(t) = x -> H(x,t)
Hx(x) = t -> H(x,t)
dH(t) = x -> ForwardDiff.derivative(y->Hx(x)(y).array,t)
dH(tv)(xv)

# Whereas everything smooth without VectorValue, i.e., arrays
K(x,t) =    [5*x[1]*x[2],x[2]^2+t]
Kx(x) = t -> K(x,t)
dK(t) = x -> ForwardDiff.derivative(y->Kx(x)(y),t)
dK(tv)(xv)

# The VectorValue treatment is OK
A(t) = VectorValue(5*t,2*t)
ForwardDiff.derivative(y->A(y).array,tv)

# As above, no problem for practically univariate functions
B(x,t) = A(t)
ForwardDiff.derivative(y->B(xv,y).array,tv)

# It also works (sure)
C(x) = t -> A(t)
ForwardDiff.derivative(y->C(xv)(y).array,tv)

# But it does not
D(x,t) = VectorValue(5*x[1]*x[2],x[2]^2+t)
ForwardDiff.derivative(y->D(xv,y).array,tv)

# Neither this (as expected)
E(x) = t -> VectorValue(5*x[1]*x[2],x[2]^2+t)
ForwardDiff.derivative(y->E(xv)(y).array,tv)

# That is surprising
xv = Point(rand(Float64,2)...)
FF = t -> VectorValue(5*xv[1]*xv[2],xv[2]^2+t)
ForwardDiff.derivative(y->FF(y).array,tv)

# Or even in a function (sth about the global scope...)
function mytry(xv,D)
    FF(t) = D(xv,t)
    ForwardDiff.derivative(y->FF(y).array,tv)
end
GG = mytry(xv,D)
```

**Comment by santiagobadia** [comment]
@fverdugo @jesusbonilla 

if you can take a look at this problem, and find a solution, let me know

**Comment by jesusbonilla** [comment]
I think it could be related with the fact that all the entries in `VectorValue` need to be of the same type. Then, when you define: 
```
H(x,t) = VectorValue(5*x[1]*x[2],x[2]^2+t)
```
Inside `ForwardDiff.derivative`, you will call `H(x::VectorValue{Float64},t::Forwardiff.Dual{Tag,Float64})`. Hence, the first entry of H would be a `Float64` while the second a `ForwardDiff.Dual`, and thus it crashes. If this is the case, then:
```
H1(x,t) = VectorValue(5*x[1]*x[2]+0*t,x[2]^2+t)
```
should work, but I am not 100% sure if this is the issue that you are facing.

**Comment by santiagobadia** [comment]
So, it cannot derive constant functions?

That is really dirty, users won't put 0*t

Any idea how to fix this issue?

**Comment by jesusbonilla** [comment]
I agree `0*t` is not the way to go, it is just a workaround to check if this was your issue. 

The problem is in the creation of the `VectorValue`. Thus, I would propose a creator for `VectorValue` that converts all entries to the same type, or a least in the case of mixing floats with duals. I can work on it, but right now I do not see an elegant straightforward solution. I will work on this.

**Comment by fverdugo** [comment]
This works:

```julia
H(x,t) = VectorValue(promote(5*x[1]*x[2],x[2]^2+t))
Ht(t) = x -> H(x,t)
Hx(x) = t -> H(x,t)
dH(t) = x -> VectorValue(ForwardDiff.derivative(y->get_array(Hx(x)(y)),t)
dH(tv)(xv)
```
I will introduce the `promotion` under the hood so that the user can write the code more naturally. See
https://github.com/gridap/Gridap.jl/issues/297

**Comment by fverdugo** [comment]
Fix on the way in https://github.com/gridap/Gridap.jl/pull/298

**Comment by santiagobadia** [comment]
Solved in #20
